Userland/Libraries/LibCore/SharedCircularQueue.h

```diff
@@ -0,0 +1,239 @@
+    {
+        return ((m_queue->m_queue->m_head.load() - 1) % Size) != (m_queue->m_queue->m_tail.load() % Size); 
+    }
+
+    // Repeatedly try to enqueue, using the wait_function to wait if it's not possible
+    ErrorOr<void> try_blocking_enqueue(ValueType to_insert, Function<void()> wait_function)
+    {
+        ErrorOr<void, QueueStatus> result;
+        while (true) {
+            result = try_enqueue(to_insert);
+
+            if (result.is_error()) {
+                if (result.error() == QueueStatus::Full) {
+                    wait_function();
+                } else {
+                    return Error::from_string_literal("Unexpected error while enqueuing"sv);
+                }
+            } else {
+                break;
+            }
+        }
+        return {};
+    }
+
+    // Roughly adopted for array queues from
+    // https://www.cs.rochester.edu/~scott/papers/1996_PODC_queues.pdf
+    // FIXME: Add proofs that this algorithm is correct under all circumstances and invariants are sustained.
+    ErrorOr<ValueType, QueueStatus> try_dequeue()
+    {
+        VERIFY(!m_queue.is_null());
+        // (1)
+        auto next = m_queue->m_queue->m_head.load();
+        // (2) Loop invariant: next always points to the value in the queue that may be returned.
+        while (true) {
+            // (3) The >= is not strictly necessary, but it feels safer :^)
+            if (next >= m_queue->m_queue->m_tail.load())
+                return QueueStatus::Empty;
+
+            // (4) This CAS only succeeds if nobody modified the head while we did the above work.
+            if (m_queue->m_queue->m_head.compare_exchange_strong(next, next + 1))
+                // (5)
+                return m_queue->m_queue->m_data[next % Size];
+        }
+    }
+
+private:
+    struct SharedMemorySPCQ {
+        explicit SharedMemorySPCQ() = default;
+        SharedMemorySPCQ(SharedMemorySPCQ const&) = delete;
+        SharedMemorySPCQ(SharedMemorySPCQ&&) = delete;
+        ~SharedMemorySPCQ() = default;
+
+        // Invariant: tail >= head
+        // Invariant: head and tail are monotonically increasing
+        // Invariant: tail always points to the next free location where an enqueue can happen.
+        // Invariant: head always points to the element to be dequeued next.
+        // Invariant: tail is only modified by enqueue functions.
+        // Invariant: head is only modified by dequeue functions.
+        // An empty queue is signalled with:  tail = head
+        // A full queue is signalled with:  head - 1 mod size = tail mod size  (i.e. head and tail point to the same index in the data array)
+        // FIXME: These invariants aren't proven to be correct after each successful completion of each operation where it is relevant.
+        //        The work could be put in but for now I think the algorithmic correctness proofs of the functions are enough.
+        Atomic<size_t, AK::MemoryOrder::memory_order_seq_cst> m_tail { 0 };
```

Comment:
You will likely want to align these to cache line size for the target architecture, so they sit on separate cache lines otherwise performance will likely suffer significantly due to [false sharing](https://en.wikipedia.org/wiki/False_sharing).

Reply:
That is an extremely smart observation, will do.